README.txt
Introduction
============

Caching helper function used in affinitic zope/plone projects



docs/HISTORY.txt
Changelog
=========

0.7.2 (unreleased)
------------------

- Add i18n support for ZPT file
  [macagua]


0.7.1 (2016-06-03)
------------------

- Testing layer that avoids caching
  [gotcha]


0.7 (2016-06-03)
----------------

- Enable/disable caching via python or browser view
  [gotcha]


0.6.2 (2016-05-25)
------------------

- Add environment variable to set default cache lifetime
  [jfroche]

- Fix dependencies storage
  [jfroche]


0.6.1 (2015-09-22)
------------------

- Use advanced dict to allow getattr on cached rowproxy refs #7577
  [schminitz]


0.6 (2014-10-22)
----------------

- Allow to invalidate dependencies


0.5 (2014-08-26)
----------------

- Move memcache utility to overrides zcml

- Delegate the namespace definition to a utility and cache the namespace calculation

- Make sqlalchemy a soft dependency


0.4 (2013-02-22)
----------------

- Optimizes the caching functions.

- Adds two decorators who clears the cache before or after a function.

- Adds a function to clear a specific cache.

- Adds a function to invalidate a specific key in the cache.


0.3 (2013-02-13)
----------------

- Change name from arsia.caching to affinitic.caching


0.2 (2011-02-22)
----------------

- Fix dependencies

0.1 (2010-11-01)
----------------

- Initial release


src/arsia/caching/tests/test_doctest.py
# -*- coding: utf-8 -*-
"""
arsia.cerise.core

Licensed under the GPL license, see LICENCE.txt for more details.
Copyright by Affinitic sprl

$Id: event.py 67630 2006-04-27 00:54:03Z jfroche $
"""
import unittest

from zope.testing import doctest

from zope.app.testing.placelesssetup import setUp, tearDown
from zope.configuration.xmlconfig import XMLConfig

optionflags = doctest.REPORT_ONLY_FIRST_FAILURE | doctest.ELLIPSIS

import arsia.caching


class B(object):
    value = None


class User(object):
    pass


def configurationSetUp(test):
    setUp()
    XMLConfig('testing.zcml', arsia.caching)()


def configurationTearDown(test):
    tearDown()

OPTIONFLAGS = (doctest.ELLIPSIS |
               doctest.REPORT_ONLY_FIRST_FAILURE |
               doctest.NORMALIZE_WHITESPACE)


def test_suite():
    tests = (
        doctest.DocFileSuite('memcached.txt',
                             optionflags=OPTIONFLAGS,
                             package="arsia.caching",
                             setUp=configurationSetUp,
                             tearDown=configurationTearDown),
        )
    return unittest.TestSuite(tests)


src/affinitic/caching/tests/test_doctest.py
# -*- coding: utf-8 -*-
"""
affinitic.cerise.core

Licensed under the GPL license, see LICENCE.txt for more details.
Copyright by Affinitic sprl

$Id: event.py 67630 2006-04-27 00:54:03Z jfroche $
"""
import unittest

from zope.testing import doctest

from zope.app.testing.placelesssetup import setUp, tearDown
from zope.configuration.xmlconfig import XMLConfig

optionflags = doctest.REPORT_ONLY_FIRST_FAILURE | doctest.ELLIPSIS

import affinitic.caching


class B(object):
    value = None


class User(object):
    pass


def configurationSetUp(test):
    setUp()
    XMLConfig('testing.zcml', affinitic.caching)()


def configurationTearDown(test):
    tearDown()

OPTIONFLAGS = (doctest.ELLIPSIS |
               doctest.REPORT_ONLY_FIRST_FAILURE |
               doctest.NORMALIZE_WHITESPACE)


def test_suite():
    tests = (
        doctest.DocFileSuite('memcached.txt',
                             optionflags=OPTIONFLAGS,
                             package="affinitic.caching",
                             setUp=configurationSetUp,
                             tearDown=configurationTearDown),
        )
    return unittest.TestSuite(tests)


src/affinitic/caching/memoize.py
# encoding: utf-8
"""
memoize.py

Created by jfroche.
Licensed under the GPL license, see LICENCE.txt for more details.
Copyright by Affinitic sprl
"""

from zope import component

from plone.memoize import volatile
from plone.memoize.interfaces import ICacheChooser

from affinitic.caching.interfaces import IInvalidateKey


def cache_for_functions(func, lifetime=86400):
    """
    Use this to decorate statics methods or separate functions.
    You can define a lifetime duration for the value that are cached with the
    lifetime parameter (in second), by default it's 1 day.
    """

    def store_in_cache(func, *args, **kwargs):
        key = '%s.%s' % (func.__module__, func.__name__)
        cache_chooser = component.getUtility(ICacheChooser)
        cache = cache_chooser(key)
        if hasattr(cache, 'client'):
            cache.client.defaultAge = lifetime
        else:
            cache.ramcache.update(lifetime)
        return cache

    def get_key(func, *args, **kwargs):
        return (args, frozenset(kwargs.items()), )

    def cache(get_key):
        return volatile.cache(get_key, get_cache=store_in_cache)

    return cache(get_key)(func)


def cache_for_instances(func, lifetime=86400):
    """
    Use this to decorate class instances methods only.
    You can define a lifetime duration for the value that are cached with the
    lifetime parameter (in second), by default it's 1 day.
    """

    def store_in_cache(func, *args, **kwargs):
        key = '%s.%s' % (func.__module__, func.__name__)
        cache_chooser = component.getUtility(ICacheChooser)
        cache = cache_chooser(key)
        if hasattr(cache, 'client'):
            cache.client.defaultAge = lifetime
        else:
            cache.ramcache.update(lifetime)
        return cache

    def get_key(func, *args, **kwargs):
        # remove 'self' attribute received from instance method
        return (args[1:], frozenset(kwargs.items()), )

    def cache(get_key):
        return volatile.cache(get_key, get_cache=store_in_cache)

    return cache(get_key)(func)


def clear_after(funcname, args=(), kwargs={}):
    """
    Use this to decorate class methods or functions.
    This decorator invalidates the given function in the cache depending of
    his arguments after the decorate function was executed
    See clear_cache docs for details on parameters.
    """

    def get_key(arg, kwargs):
        return (arg, frozenset(kwargs.items()), )

    def clear(func):

        def wrapped_clear(*args, **kwargs):
            func(*args, **kwargs)
            component.queryUtility(IInvalidateKey)(funcname, key)
        return wrapped_clear
    key = '%s:%s' % (funcname, get_key(args, kwargs))
    return clear


def clear_before(funcname, args=(), kwargs={}):
    """
    Use this to decorate class methods or functions.
    This decorator invalidates the given function in the cache depending of
    his arguments before the decorate function was executed
    See clear_cache docs for details on parameters.
    """

    def get_key(arg, kwargs):
        return (arg, frozenset(kwargs.items()), )

    def clear(func):

        def wrapped_clear(*args, **kwargs):
            component.queryUtility(IInvalidateKey)(funcname, key)
            func(*args, **kwargs)
        return wrapped_clear
    key = '%s:%s' % (funcname, get_key(args, kwargs))
    return clear


def clear_cache(funcname, args=(), kwargs={}):
    """
    Use this function to invalidate a specific function or methods depending
    of his arguments in the cache
    Parameters : - funcname : a string with the complete path and the function
                              name. ex: 'arsia.cerise.core.memoize.clear_cache'
                 - args : a tuple with the arguments passed to the function.
                          ex: ('param1', )
                 - kwargs : a dictionary with the keyword arguments passed
                            to the function. ex: {'key1': value}
    """

    def get_key(arg, kwargs):
        return (arg, frozenset(kwargs.items()), )

    key = '%s:%s' % (funcname, get_key(args, kwargs))
    component.queryUtility(IInvalidateKey)(funcname, key)


def invalidate_key(funcname, key):
    cache = component.queryUtility(ICacheChooser)(key)
    cache.ramcache.invalidate(funcname, key=dict(key=cache._make_key(key)))


def one_day_memoize_for_instances(func):
    """
    Use this to decorate class instances methods only
    Cached return values are cleared every day
    """
    return cache_for_instances(func, lifetime=86400)


def one_day_memoize_for_functions(func):
    """
    Use this to decorate statics methods or separate functions.
    Cached return values are cleared every day.
    """
    return cache_for_functions(func, lifetime=86400)


def one_hour_memoize_for_instances(func):
    """
    Use this to decorate class instances methods only
    Cached return values are cleared every hour
    """
    return cache_for_instances(func, lifetime=3600)


def one_hour_memoize_for_functions(func):
    """
    Use this to decorate statics methods or separate functions.
    Cached return values are cleared every hour
    """
    return cache_for_functions(func, lifetime=3600)


src/affinitic/caching/memcached.py
# -*- coding: utf-8 -*-
"""
affinitic.caching

Licensed under the GPL license, see LICENCE.txt for more details.
Copyright by Affinitic sprl

$Id: event.py 67630 2006-04-27 00:54:03Z jfroche $
"""
import cPickle
import md5
import os
import pkg_resources

try:
    pkg_resources.get_distribution('sqlalchemy')
except pkg_resources.DistributionNotFound:

    class RowProxy(object):
        pass
else:  # we have sqlalchemy
    try:
        from sqlalchemy.engine.base import RowProxy
    except ImportError, e:
        from sqlalchemy.engine import RowProxy


from lovely.memcached.event import invalidateCache
from lovely.memcached.event import InvalidateCacheEvent
from lovely.memcached.interfaces import IMemcachedClient
from lovely.memcached.utility import MemcachedClient

from zope import component
from zope import event
from zope.component import queryUtility
from zope.ramcache.interfaces.ram import IRAMCache

from plone.memoize import volatile
from plone.memoize.interfaces import ICacheChooser
from plone.memoize.ram import (AbstractDict, store_in_cache, RAMCacheAdapter)

from affinitic.caching.interfaces import IMemcachedDefaultNameSpace


class MemcachedClientWithNameSpace(MemcachedClient):
    """
    Delegate the namespace definition to a utility
    Cache the namespace calculation
    """
    _defaultNS = None

    @property
    def defaultNS(self):
        if self._defaultNS is not None:
            return self._defaultNS
        defaultNameSpace = queryUtility(IMemcachedDefaultNameSpace)
        if defaultNameSpace is not None:
            defaultNameSpace = defaultNameSpace()
            if defaultNameSpace is not None:
                self._defaultNS = defaultNameSpace

    def _getNS(self, ns, raw):
        defaultNameSpace = self.defaultNS or 'defaultNS'
        if not ns and defaultNameSpace:
            if raw:
                ns = str(defaultNameSpace)
            else:
                ns = defaultNameSpace
        return ns or None


def memcachedClient():
    servers = os.environ.get("MEMCACHE_SERVER", "127.0.0.1:11211").split(",")
    defaultAge = int(os.environ.get("MEMCACHE_DEFAULT_AGE", '86400'))
    return MemcachedClientWithNameSpace(servers, defaultNS=None,
                                        defaultAge=defaultAge)


class MemcacheAdapter(AbstractDict):

    def __init__(self, client, globalkey=''):
        self.client = client

    def _make_key(self, source):
        return md5.new(source).hexdigest()

    def __getitem__(self, key):
        cached_value = self.client.query(self._make_key(key), raw=True)
        if cached_value is None:
            raise KeyError(key)
        else:
            return cPickle.loads(cached_value)

    def __setitem__(self, key, value):
        try:
            cached_value = cPickle.dumps(value)
        except TypeError:
            if isinstance(value, list) and isinstance(value[0], RowProxy):
                value = [RowProxyDict(d) for d in value]
            elif isinstance(value, RowProxy):
                value = RowProxyDict(value)
            else:
                raise
            cached_value = cPickle.dumps(value)
        self.client.set(cached_value, self._make_key(key), raw=True,
                        dependencies=[])

    def set(self, key, value, dependencies=[]):
        try:
            cached_value = cPickle.dumps(value)
        except TypeError:
            if isinstance(value, list) and isinstance(value[0], RowProxy):
                value = [RowProxyDict(d) for d in value]
            elif isinstance(value, RowProxy):
                value = RowProxyDict(value)
            else:
                raise
            cached_value = cPickle.dumps(value)
        self.client.set(cached_value, self._make_key(key), raw=True,
                        dependencies=dependencies)

    def setWithLifetime(self, key, value, lifetime, dependencies=[]):
        cached_value = cPickle.dumps(value)
        self.client.set(cached_value, self._make_key(key), raw=True,
                        lifetime=lifetime,
                        dependencies=dependencies)


def choose_cache(fun_name):
    client = queryUtility(IMemcachedClient)
    if client is not None:
        return MemcacheAdapter(client, globalkey=fun_name)
    else:
        return RAMCacheAdapter(queryUtility(IRAMCache),
                               globalkey=fun_name)


_marker = object()


def cache(get_key, dependencies=None, get_dependencies=None, lifetime=None):

    def decorator(fun):

        def replacement(*args, **kwargs):
            try:
                key = get_key(fun, *args, **kwargs)
            except volatile.DontCache:
                return fun(*args, **kwargs)
            key = str(key)

            # Do not cache when not using memcache with dependencies
            memcache_client = queryUtility(IMemcachedClient)
            if (dependencies is not None or get_dependencies is not None) and not memcache_client:
                return fun(*args, **kwargs)

            if dependencies is not None or get_dependencies is not None:
                deps = dependencies
                if get_dependencies is not None:
                    deps = get_dependencies(fun, *args, **kwargs)
            cache = store_in_cache(fun, *args, **kwargs)
            cached_value = cache.get(key, _marker)
            if cached_value is _marker:
                if lifetime is None:
                    cached_value = fun(*args, **kwargs)
                    cache.set(key, cached_value, deps)
                else:
                    cached_value = fun(*args, **kwargs)
                    cache.setWithLifetime(key, cached_value, lifetime, deps)
            return cached_value
        return replacement
    return decorator


def invalidate_key(funcname, key):
    client = queryUtility(IMemcachedClient)
    cache = queryUtility(ICacheChooser)(key)
    if client is not None:
        invalidateEvent = InvalidateCacheEvent(key=cache._make_key(key),
                                               raw=True)
        event.notify(invalidateEvent)
    else:
        key = dict(key=cache._make_key(key))
        cache.ramcache.invalidate(funcname, key=key)


def invalidate_dependencies(dependencies):
    """
    Invalidate all caches linked to dependencies
    """
    client = queryUtility(IMemcachedClient)
    component.provideHandler(invalidateCache)
    # memcached
    if client is not None:
        invalidateEvent = InvalidateCacheEvent(raw=True,
                                               dependencies=dependencies)
        event.notify(invalidateEvent)
    # ramcache
    else:
        # Cannot invalidate dependencies with ramcache
        pass


class RowProxyDict(dict):
    """
    dict allowing to get values via __getattr__
    """

    def __getattr__(self, item):
        """Maps values to attributes.
        Only called if there *isn't* an attribute with this name
        """
        if self.has_key(item):
            return self.__getitem__(item)
        elif self.has_key(item.upper()):
            return self.__getitem__(item.upper())
        else:
            raise AttributeError(item)
